File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */


class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // stack<int> stackOne;
        // stack<int> stackTwo;

        // ListNode* node = l1;
        // while(node != nullptr) { 
        //     stackOne.push(node->val);
        //     node = node->next;
        // }

        // node = l2;
        // while(node != nullptr) { 
        //     stackTwo.push(node->val);
        //     node = node->next;
        // }

        // int numOne = 0;
        // int i = 1;
        // while(!stackOne.empty()) {
        //     numOne +=  stackOne.pop() * i;
        //     i *= 10;
        // }

        // int numTwo = 0;
        // i = 1;
        // while(!stackTwo.empty()) {
        //     numTwo +=  stackTwo.pop() * i;
        //     i *= 10;
        // }

        // int result = numOne + numTwo;

        bool carryOver = false;
        int currNum = 0;
        // bool first = true;
        stack<ListNode*> nodeStack;

        ListNode* currOne  = l1;
        ListNode* currTwo = l2;

        while(currOne || currTwo) {
            ListNode* node = new ListNode;

            // currNum = (stackOne.empty()) : stackTwo.pop() ? stackOne.pop() + stackTwo.pop();

            if (!currTwo) {
                currNum = currOne->val;
                currOne = currOne->next;
            } else if (!currOne) {
                currNum = currTwo->val;
                currTwo = currTwo->next;
            } else {
                currNum = currOne->val + currTwo->val;
                currOne = currOne->next;
                currTwo = currTwo->next;
            }
            
            if (carryOver) {
                currNum += 1;
                carryOver = false;
            }


            if (currNum > 9) {
                carryOver = true;
                currNum -= 10;
            }

            ListNode* newNode = new ListNode(currNum);
            nodeStack.push(newNode);
        }

        // Deal with excess carry over
        if (carryOver) {
            ListNode* newNode = new ListNode(1);
            nodeStack.push(newNode);
        }

        ListNode* root = nullptr;
        while (nodeStack.empty() == false) {
            ListNode* newNode = nodeStack.top();
            nodeStack.pop();
            newNode->next = root;
            root = newNode;
        }

        return root;

    }
};

File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        head = ListNode() # Dummy Head
        curr = head
        node = None

        l1_curr = l1
        l2_curr = l2

        carry = False

        while l1_curr or l2_curr:
            if l1_curr and l2_curr:
                sum = l1_curr.val + l2_curr.val + carry
                l1_curr = l1_curr.next
                l2_curr = l2_curr.next
            elif l1_curr:
                sum = l1_curr.val + carry
                l1_curr = l1_curr.next 
            else:
                sum = l2_curr.val + carry
                l2_curr = l2_curr.next
                        
            if sum >= 10:
                node = ListNode(sum % 10)
                carry = True
            else:
                node = ListNode(sum)
                carry = False
            
            curr.next = node
            curr = curr.next
        
        if carry:
            curr.next = ListNode(1)
        
        return head.next


